let buttons template accept string replacement from existing slot-values

8key opened the pull request:
**Proposed changes**:
- as an example in domain.yml, if we define buttons:
```templates:
  utter_ask_example:
    - text: "example {slot_1}？"
      buttons:
      - title: "{slot_2}"
        payload: '/choose{"slot_2": "{slot_2}"}'
```
during the templating, only the wildcard {slot_1} will be filled with the value of slot_1, wildcards defined in buttons will not be processed. 

This Pull Request is to update with adding a feature such that wildcards defined in the buttons template will be picked and replaced. This is particularly useful when button slot values are dynamic and generated as conversation flows. 

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=5850) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=5850) before we can accept your contribution.<br/><hr/>**8key** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=5850) it.</sub>

sara-tagger commented:
Thanks for submitting a pull request 🚀 @JustinaPetr will take a look at it as soon as possible ✨

8key commented:
Yep, let me take a look first.

Best



________________________________
From: Tobias Wochinger <notifications@github.com>
Sent: Tuesday, May 26, 2020 5:06 PM
To: RasaHQ/rasa <rasa@noreply.github.com>
Cc: Greg Dingyu Zhang <t_zhangdingyu@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [RasaHQ/rasa] let buttons template accept string replacement from existing slot-values (#5850)


@wochinge requested changes on this pull request.

@8key<https://github.com/8key> Thanks a lot for your contribution and your initiative to fix this! 🎉Funny enough, this is actually already supported but doesn't work due to a bug.

It should be buttons (note the s) and not button in this line https://github.com/RasaHQ/rasa/blob/96a46713869b7ef7dd40885d1d173f6d97d07bfe/rasa/core/nlg/template.py#L112 .

This would also require fixing some of the tests in the following module, e.g.: https://github.com/RasaHQ/rasa/blob/39b15969935c619a5e6f7298fe4577799ef4286b/tests/core/test_nlg.py#L266 (all the tests use button instead of buttons).

Would you mind adapting your pull request? 🙏

It might further make sense to introduce a constant for buttons to avoid this in the future. However, it might make sense to handle this in a separate PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/RasaHQ/rasa/pull/5850#pullrequestreview-418108193>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHAN6ZHJWRKUYMXGAB2YKFTRTOBCJANCNFSM4NEIKHJA>.

8key commented:
> @8key Thanks a lot for your contribution and your initiative to fix this! 🎉Funny enough, this is actually already supported but doesn't work due to a bug.
> 
> It should be `buttons` (note the `s`) and not `button` in this line
> 
> https://github.com/RasaHQ/rasa/blob/96a46713869b7ef7dd40885d1d173f6d97d07bfe/rasa/core/nlg/template.py#L112
> 
> .
> This would also require fixing some of the tests in the following module, e.g.:
> 
> https://github.com/RasaHQ/rasa/blob/39b15969935c619a5e6f7298fe4577799ef4286b/tests/core/test_nlg.py#L266
> 
> (all the tests use `button` instead of `buttons`).
> Would you mind adapting your pull request? 🙏
> 
> It might further make sense to introduce a constant for `buttons` to avoid this in the future. However, it might make sense to handle this in a separate PR.

@wochinge Just had a second check, simply fixing `button` --> `buttons` won't fix the bug, as the input to `interpolate ` function
https://github.com/RasaHQ/rasa/blob/96a46713869b7ef7dd40885d1d173f6d97d07bfe/rasa/core/nlg/interpolator.py#L42
is either `dict` or `str`, whereas the input from `buttons` is `list`. instead of touching that recursive `interpolate` function, I added a new function to handle `buttons` wildcard replacement. 

Can you take a look and let me know. 

Thanks

wochinge commented:
@8key Is this ready for another review?

8key commented:
Yes

________________________________
From: Tobias Wochinger <notifications@github.com>
Sent: Wednesday, June 3, 2020 4:19 PM
To: RasaHQ/rasa <rasa@noreply.github.com>
Cc: Greg Dingyu Zhang <t_zhangdingyu@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [RasaHQ/rasa] let buttons template accept string replacement from existing slot-values (#5850)


@8key<https://github.com/8key> Is this ready for another review?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/RasaHQ/rasa/pull/5850#issuecomment-638040648>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHAN6ZH2ZIRYKRSN357XBGTRUYBQVANCNFSM4NEIKHJA>.

wochinge commented:
Awesome, will give it a review as soon as possible 👍

wochinge commented on rasa/core/nlg/template.py:
```diff
@@ -110,13 +110,17 @@ def _fill_template(
             "image",
             "custom",
             "button",
+            "buttons",
```
```suggestion
            "buttons",
```

wochinge replied:
that should be the only change you need (+ fixing the test cases)

8key replied:
Hello, Have done these

8key replied:
anything else?

wochinge commented on rasa/core/nlg/template.py:
```diff
@@ -109,14 +113,19 @@ def _fill_template(
             for key in keys_to_interpolate:
                 if key in template:
-                    template[key] = interpolate(template[key], template_vars)
+                    if key != "buttons":
+                        template[key] = interpolate(template[key], template_vars)
+                    else:
+                        template[key] = interpolate_buttons(
+                            template[key], template_vars
+                        )
```
that's not needed, right?

8key replied:
These are needed, the original code won't work otherwise.

wochinge commented on rasa/core/nlg/interpolator.py:
```diff
@@ -53,3 +53,23 @@ def interpolate(
                 template[k] = interpolate_text(v, values)
         return template
     return template
+
+
+def interpolate_button_text(template: Text, values: Dict[Text, Text]) -> Text:
+    for slot in values:
+        if values[slot]:
+            template = template.replace("{" + str(slot) + "}", values[slot])
+    return template
+
+
+def interpolate_buttons(template: List[Dict], values: Dict[Text, Text]) -> List[Dict]:
+    if isinstance(template, list):
+        for idx, a_button_dict in enumerate(template):
+            if isinstance(a_button_dict, dict):
+                template[idx]["payload"] = interpolate_button_text(
+                    a_button_dict["payload"], values
+                )
+                template[idx]["title"] = interpolate_button_text(
+                    a_button_dict["title"], values
+                )
+    return template
```
that's not needed, right?

8key replied:
These are needed, the original code won't work otherwise.

wochinge replied:
@8key Got the problem now. I think we should handle this more general and not just for buttons. The problem is that we are currently interpolating `dict` and `string` items, but no `list`s. 

This should fix it for all sorts of keys:

```python
def interpolate(
    template: Union[List[Any], Dict[Text, Any], Text], values: Dict[Text, Text]
) -> Union[List[Any], Dict[Text, Any], Text]:
    if isinstance(template, str):
        return interpolate_text(template, values)
    elif isinstance(template, dict):
        for k, v in template.items():
            if isinstance(v, dict):
                interpolate(v, values)
            elif isinstance(v, list):
                template[k] = [interpolate(i, values) for i in v]
            else:
                template[k] = interpolate_text(v, values)
        return template
    # This is the added the case which we missed before
    elif isinstance(template, list):   
        return [interpolate(i, values) for i in template]

    return template
```

8key replied:
Okay, make sense, will update thanks

wochinge replied:
Thanks for fixing this 🚀 🙏

8key replied:
Added this line, would have a missing key error otherwise.
this is because
button template has value e.g., 
`/choose{"appt_time": "{candidate_appt_time_1}"}`
need it to be
`/choose{{"appt_time": "{candidate_appt_time_1}"}}`
that the outer curly braces have to be escaped while apply `.format` method
https://github.com/8key/rasa/blob/7773e40427c251a0b762f61314fa8cc6cf169675/rasa/core/nlg/interpolator.py#L25

8key replied:
let me know if anything

wochinge commented on rasa/core/nlg/interpolator.py:
```diff
@@ -15,6 +15,15 @@ def interpolate_text(template: Text, values: Dict[Text, Text]) -> Text:
     # (b) { or } in slot name
     try:
         text = re.sub(r"{([^\n{}]+?)}", r"{0[\1]}", template)
+        # for button template, we need to handle escapable curly brace, e.g., button template has
```
@8key Tough problem 😬 Actually you can workaround this by double escaping the entity.

See this example
```
def test_interpolate_text():
    text = '/greet{{"entity": "{test}"}}'
    values = {"test": "hello"}
    result = interpolate_text(text, values)

    assert result == '/greet{"entity": "hello"}'
```

8key replied:
@wochinge That's why i had added line 26 to handle this. I trust it's good to merge

wochinge replied:
Sorry, I think my message wasn't particularly clear:

You don't need the change in line 26 as you can double escape things which shouldn't be interpolated in your `domain.yml` by double escaping them, e.g.

```
responses:
  utter_greet:
  - buttons:
    - payload: '/greet{{"some_entity": "{test}"}}'
      title: Hi {test}
```

wochinge commented on tests/core/test_nlg.py:
```diff
@@ -260,15 +260,31 @@ def test_nlg_fill_template_attachment(attach_slot_name, attach_slot_value):
+    template = {
+        "buttons": [
+            {
+                "payload": '/choose{{{{"some_slot": "{{{}}}"}}}}'.format(
+                    button_slot_name
+                ),
```
```suggestion
                "payload": f'/choose{{{{"some_slot": "{{{button_slot_name}}}"}}}}',
```

wochinge commented on tests/core/test_nlg.py:
```diff
@@ -260,15 +260,31 @@ def test_nlg_fill_template_attachment(attach_slot_name, attach_slot_value):
+                "payload": '/choose{{{{"some_slot": "{{{}}}"}}}}'.format(
+                    button_slot_name
+                ),
+                "title": "{{{}}}".format(button_slot_name),
```
```suggestion
                "title": f"{{{button_slot_name}}}",
```

wochinge commented on tests/core/test_nlg.py:
```diff
@@ -260,15 +260,31 @@ def test_nlg_fill_template_attachment(attach_slot_name, attach_slot_value):
+    assert result == {
+        "buttons": [
+            {
+                "payload": '/choose{{"some_slot": "{}"}}'.format(button_slot_value),
```
```suggestion
                "payload": f'/choose{{"some_slot": "{button_slot_value}"}}',
```

wochinge commented on tests/core/test_nlg.py:
```diff
@@ -260,15 +260,31 @@ def test_nlg_fill_template_attachment(attach_slot_name, attach_slot_value):
+        "buttons": [
+            {
+                "payload": '/choose{{"some_slot": "{}"}}'.format(button_slot_value),
+                "title": "{}".format(button_slot_value),
```
```suggestion
                "title": f"{button_slot_value}"
```

wochinge commented on rasa/core/nlg/interpolator.py:
```diff
@@ -39,8 +39,8 @@ def interpolate_text(template: Text, values: Dict[Text, Text]) -> Text:
-    template: Union[Dict[Text, Any], Text], values: Dict[Text, Text]
-) -> Union[Dict[Text, Any], Text]:
+    template: Union[List[Dict], Dict[Text, Any], Text], values: Dict[Text, Text]
+) -> Union[List[Dict], Dict[Text, Any], Text, None]:
```
how should it return `None`?

8key commented:
@wochinge Updated accordingly, pls check

wochinge commented on rasa/core/nlg/interpolator.py:
```diff
@@ -52,4 +52,6 @@ def interpolate(
-    return template
+    elif isinstance(template, list):
+        return [interpolate(i, values) for i in template]
+    return interpolate_text(template, values)
```
Arriving at this point means that we don't know how to handle the type of `template`. I in this case we should just hold off and return it as it is. 
```suggestion
    return template
```

wochinge commented on changelog/5850.bugfix.rst:
```diff
@@ -0,0 +1,3 @@
+Slots will be correctly interpolated for buttons template.
```
```suggestion
Slots will be correctly interpolated for `button` responses.
```

8key replied:
hello,

i have checked in

wochinge commented:
@8key Can you please sign the Contributor License Agreement so that we can merge? 🙏

8key commented:
> @8key Can you please sign the Contributor License Agreement so that we can merge? 🙏

Hello @wochinge I am not sure if i have signed, i only saw this

![license](https://user-images.githubusercontent.com/29417316/89241716-a5821880-d5cd-11ea-837c-62e80ab5556a.JPG)

wochinge commented:
> Hello @wochinge I am not sure if i have signed, i only saw this

Exactly this one. If you scroll down, there should be a button to sign it.

8key commented:
> > Hello @wochinge I am not sure if i have signed, i only saw this
> 
> Exactly this one. If you scroll down, there should be a button to sign it.

But it says I have already signed. 

![Screen Shot 2020-08-04 at 9 57 26 AM](https://user-images.githubusercontent.com/29417316/89302533-f2e79f80-d638-11ea-9add-5193d0cad50f.png)

8key commented:
> > > Hello @wochinge I am not sure if i have signed, i only saw this
> > 
> > 
> > Exactly this one. If you scroll down, there should be a button to sign it.
> 
> But it says I have already signed.
> 
> ![Screen Shot 2020-08-04 at 9 57 26 AM](https://user-images.githubusercontent.com/29417316/89302533-f2e79f80-d638-11ea-9add-5193d0cad50f.png)

@wochinge could that be because the previous commit used my local user account which is greg@dingyus-MacBook-pro.local, do you know how to fix this?

wochinge commented:
> could that be because the previous commit used my local user account which is greg@dingyus-MacBook-pro.local, do you know how to fix this?

@8key Yes, I think that's it. Some commits were made with an Git author which isn't connected to GitHub. You could do rebase which squashes all commits. You should just make sure that the rebase is done by the Git author which is connected to your GitHub account. `git config --global user.email` should be an email which is [connected to your GitHub account](https://github.com/settings/emails). If you used a different email for the other account, you could also try adding this one to GitHub 🤔

8key commented:
> > could that be because the previous commit used my local user account which is [greg@dingyus-MacBook-pro.local](mailto:greg@dingyus-MacBook-pro.local), do you know how to fix this?
> 
> @8key Yes, I think that's it. Some commits were made with an Git author which isn't connected to GitHub. You could do rebase which squashes all commits. You should just make sure that the rebase is done by the Git author which is connected to your GitHub account. `git config --global user.email` should be an email which is [connected to your GitHub account](https://github.com/settings/emails). If you used a different email for the other account, you could also try adding this one to GitHub 🤔

Okay let me try

wochinge commented:
@8key how is it going?

8key commented:
> @8key how is it going?

I tried to following
i am on my branch
i do rebase
```git rebase -i d4230c5```
and i change to edit
```
edit 5d8ae597f4e let button template accept string replacement from existing slot-values
pick 681b6424b04 fix format
pick 199e780222f add line space as per lint
pick 1cc86bbea5b create plotting utils
pick 279f6abc500 write errors and successes for responses
pick 82d6e68d3c3 fix file names
......
```
and i do update
```
git commit --amend --author="8key <t_zhangdingyu@hotmail.com>"
[detached HEAD c2acf72d2b6] let button template accept string replacement from existing slot-values
 Date: Mon May 18 12:44:03 2020 -0400
 2 files changed, 25 insertions(+), 3 deletions(-)
```
and i do continue
```
git rebase --continue
CONFLICT (add/add): Merge conflict in rasa/utils/plotting.py
Auto-merging rasa/utils/plotting.py
Auto-merging rasa/nlu/test.py
CONFLICT (content): Merge conflict in rasa/nlu/test.py
error: could not apply 1cc86bbea5b... create plotting utils
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 1cc86bbea5b... create plotting utils
```

which gave me conflict i don't know how to fix

Any resolution?

wochinge commented:
Mhm, it seems correct to me what you're doing.
What exactly are the merge conflicts you're getting? Did maybe accidental changes make it in the edited commit?

8key commented:
> Mhm, it seems correct to me what you're doing.
> What exactly are the merge conflicts you're getting? Did maybe accidental changes make it in the edited commit?

I followed this one
https://docs.github.com/en/github/using-git/changing-author-info
and it indeed updated names, however, i see that it took over 10k+ commits to merge, not a good sign :( 
Can you pls check. The test failed, but the license signature is passed.

8key commented:
> Mhm, it seems correct to me what you're doing.
> What exactly are the merge conflicts you're getting? Did maybe accidental changes make it in the edited commit?

Looks like the name change updated all history commits. I am not sure what to do at this stage.

8key commented:
> Mhm, it seems correct to me what you're doing.
> What exactly are the merge conflicts you're getting? Did maybe accidental changes make it in the edited commit?


On my local
```
~/workspace/rasa$ git branch
* let_buttons_template_take_value_from_current_slot
  master
~/workspace/rasa$ git log
commit f41a57ebfb27330cb11d934b3741d1f3522aa615 (HEAD -> let_buttons_template_take_value_from_current_slot, origin/let_buttons_template_take_value_from_current_slot)
Merge: d391fdefe3d 52682b3f07e
Author: Tobias Wochinger <t.wochinger@rasa.com>
Date:   Tue Aug 4 09:18:31 2020 +0200

    Merge branch 'master' into let_buttons_template_take_value_from_current_slot

commit 52682b3f07e22ab470ee3be324e088b522838da5
Merge: 13ecec48c25 28e5b7ffc87
Author: Maxime Vdb <m.verger@rasa.com>
Date:   Tue Aug 4 09:07:15 2020 +0200

    Merge pull request #6327 from RasaHQ/update-readme-to-run-docs

    update README instructions for new docs system
```

wochinge commented:
@8key I think I fixed it as I luckily had an old state on my machine 😆 I think your `amend` also changed commits which were parts of merge commits.

8key commented:
> @8key I think I fixed it as I luckily had an old state on my machine 😆 I think your `amend` also changed commits which were parts of merge commits.

Great! That's wonderful! 
but there is an error in the CI process

wochinge commented:
> but there is an error in the CI process

Mhm, that literally can't be caused by us 🙈 Will merge in latest master and see if it helps.

wochinge commented:
Merged @8key 🎉

8key commented:
> Merged @8key 🎉

Cool! @wochinge Thanks for the help

If the project has any open urgent issue that needs help, i would love to take on.

wochinge commented:
🚀 We regularly add issues to the [contributor board](https://github.com/orgs/RasaHQ/projects/23) if you want some more 🙂

8key commented:
> 🚀 We regularly add issues to the [contributor board](https://github.com/orgs/RasaHQ/projects/23) if you want some more 🙂

Got it 🙂
